### PR TITLE
storage: connection improvements

### PR DESCRIPTION
**Pull request by sploiselle**
When introducing `into_inline_connection`, the PR did not include either connections on sinks or CSR connections. This PR fixes this, in addition to propagating connection options specified on the source to the connection.

### Motivation

- This PR refactors existing code.
- This PR fixes a previously unreported bug. We did not propagate Kafka source connection options onto the source connection (though we did use it in purification).

### Tips for reviewer

Likely commit by commit.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fix a bug where options specified on Kafka sources were not propagated to the Kafka client that ingested data from the source's broker.

### CODE REVIEW

File: src/sql/src/plan/statement/ddl.rs [hunk 1]
```diff
@@ -2461,9 +2459,19 @@ fn kafka_sink_builder(
     sink_from: GlobalId,
 ) -> Result<StorageSinkConnection<ReferencedConnection>, PlanError> {
     let item = scx.get_item_by_resolved_name(&connection)?;
-    // Get Kafka connection
-    let mut connection = match item.connection()? {
-        Connection::Kafka(connection) => connection.clone(),
+    // Get Kafka connection + progress topic
+    //
+    // TODO: In ALTER CONNECTION, the progress topic must be immutable
```
Comment (jkosh44) [inline src/sql/src/plan/statement/ddl.rs:2464]: Is there an existing issue for this TODO?

**Review by jkosh44 — APPROVED** [review]
Adapter bits LGTM (I didn't look at storage parts)

File: src/sql/src/plan/statement/ddl.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:2464]: No, it's just something that exists in an unmerged PR: https://github.com/MaterializeInc/materialize/pull/21109/commits/a3e8fd40ac1a066ce95307727fc2f77d78c10b5c

Will make sure to remove the TODO.

**Comment by sploiselle** [comment]
@MaterializeInc/surfaces @MaterializeInc/testing This change had a very annoying side effect that re-opening the catalog in testdrive for the "extra tests" (where we check catalog consistency) changed the environment ID––a consequence of which is that the default progress topic changes, which meant that re-planning a `CREATE SINK` statement also gave the sink a non-equal progress topic. This then failed the catalog consistency checks.

@ParkMyCar and I chatted about this and the least invasive fix was to simply ensure that testdrive become aware of the environment ID that the tests use. I cheated a little bit and just plumbed the default environment ID everywhere because it was sufficient. If we needed more exotic plumbing of this value, we could amend this at some later point in time.

wdyat?

**Comment by philip-stoev** [comment]
@def-  has graciously agreed to work on connection-related PRs on behalf of the QA team.

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -602,13 +602,12 @@ pub fn plan_create_source(
             key: _,
         }) => {
             let connection_item = scx.get_item_by_resolved_name(connection_name)?;
-            let mut kafka_connection = match connection_item.connection()? {
-                Connection::Kafka(connection) => connection.clone(),
-                _ => sql_bail!(
+            if !matches!(connection_item.connection()?, Connection::Kafka(_)) {
+                sql_bail!(
```
Comment (def-) [inline src/sql/src/plan/statement/ddl.rs:606]: Uncovered, I tried adding a test for this:
```diff
diff --git a/test/sqllogictest/name_resolution.slt b/test/sqllogictest/name_resolution.slt
index a5fd64ba3b..08294d47f3 100644
--- a/test/sqllogictest/name_resolution.slt
+++ b/test/sqllogictest/name_resolution.slt
@@ -64,3 +64,9 @@ CREATE CONNECTION pg TO POSTGRES (
   DATABASE postgres,
   USER postgres
 ) WITH (VALIDATE = false)
+
+statement ok
+CREATE TABLE not_a_connection (a int)
+
+statement error not_a_connection is not a kafka connection
+CREATE SOURCE pg_source FROM KAFKA CONNECTION not_a_connection (TOPIC 'topic') FORMAT TEXT ENVELOPE NONE
```
But get a different error, with unresolved db/schema name btw:
```
PlanFailure:test/sqllogictest/name_resolution.slt:72:
db error: ERROR: connection "u2.u6.not_a_connection" does not exist: ERROR: connection "u2.u6.not_a_connection" does not exist
```
Am I misunderstanding how to trigger this part of the code?

File: src/storage-types/src/connections.rs [hunk 3]
```diff
@@ -243,6 +243,13 @@ impl Connection<InlinedConnection> {
             o => unreachable!("{o:?} is not an SSH connection"),
         }
     }
+
+    pub fn unwrap_csr(self) -> <InlinedConnection as ConnectionAccess>::Csr {
+        match self {
+            Self::Csr(conn) => conn,
+            o => unreachable!("{o:?} is not a Kafka connection"),
```
Comment (def-) [inline src/storage-types/src/connections.rs:250]: Uncovered, but unreachable, so fine.

**Review by def- — APPROVED** [review]
The td changes seem fine to me. Is this PR supposed to have any changes in behavior? I assume not, since no tests were touched. Nightly looks fine: https://buildkite.com/materialize/nightlies/builds/5013#_
Coverage reported inline, looks mostly fine: https://buildkite.com/materialize/coverage/builds/299

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:606]: Yeah, this checks that the connection is a Kafka connection; if it isn't a connection you get the error you saw. Pushed up a commit that tests this. However, if you use a non-Kafka connection you get an error during purification (as opposed to in planning, which is where this code lives). To trigger this error you would have to have created a catalog in a bad state--i.e. this error is unreachable/purely defensive.

**Pull request merged**